### PR TITLE
Change Table and Column meta to dict and enforce YAML output as ordered map

### PR DESCRIPTION
**Pull request by taldcroft**
### Description

This builds on top of #16240 (which built from #15496) to work toward removing `OrderedDict` from astropy where possible. This PR does so in a number of places.

#### Table

- `Table.meta` now defaults to `dict`.
- `Column.meta` now defaults to `dict`.

It is important to remember that these attributes can still be set to any mapping type (e.g. `t.meta = OrderedDict(...)`), it is just the default type that is changed.

#### YAML / ECSV

The machinery which helps to serialize `Table` and `Column` to YAML (and back) has been updated to *always* write the Table and Column `meta` attribute as an ordered map (`!!omap` in YAML). This change addresses a previous inconsistency: if the user initialized a Table like `t = Table(data, meta={'a': 1})` then the YAML would serialize `meta` as an unordered mapping. This was not the intent. The same held for `Column.meta`. Now the behavior is consistent and this is reflected in changes to unit tests and docstrings. However, the order of items in the dict is not changed.

This mostly just impacts ECSV. There should not be any issues with back-compatibility. The main change is just following the `Table` change, namely that reading an ECSV file will always return a Table that has `dict` meta instead of `OrderedDict` meta (except of course that past behavior was inconsistent).

#### Other `io.ascii` reader classes

These are somewhat unrelated to the more impacting changes but are useful nonetheless.

This PR should be squashed.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

cc: @MridulS @nstarman

### CODE REVIEW

File: astropy/cosmology/_io/ecsv.py
Comment (nstarman) [inline astropy/cosmology/_io/ecsv.py:1]: Cosmology changes LGTM. This is reversing the effects of #15484 which had updated `meta` to be a dict in the table, resulting in the change. This PR effects Table and thus Cosmology.

Comment (taldcroft) [reply astropy/cosmology/_io/ecsv.py:1]: Thanks for the review. So this is making a little churn but should be OK.

**Review by mhvk — COMMENTED** [review]
I like this approach! It puts the use of `OrderedDict` exactly where the problem lies, with yaml, and makes everything else easier. It is especially nice that order is now guaranteed to round-trip.

I assume it is on purpose that you do not (yet) change the default for `meta` itself?

**Comment by taldcroft** [comment]
> I assume it is on purpose that you do not (yet) change the default for meta itself?

Yes, but that's a great question. I was thinking about making that change as a separate PR. And as usual when you start poking at some code you see little problems. E.g. for some reason Quantity info is not enforcing that `meta` should be dict-like:
```
In [2]: q = [1, 2] * u.m

In [3]: q.info.meta = "wrong"

In [4]: q.info.meta
Out[4]: 'wrong'
```

So anyway, maybe a follow-up PR that changes the default and fixes things like that?

**Comment by mhvk** [comment]
OK, let's just deal with getting YAML right here!

**Review by mhvk — COMMENTED** [review]
Had a look through everything again and basically this looks all great to me. Probably good to have a changelog entry!

**Comment by taldcroft** [comment]
@mhvk - I've added a changelog entries and a What's New. The expectation is that the What's New will be expanded to include the next PR to change the default to `dict` in `MetaData` itself.

AFAIK this PR is complete and ready for review.

File: astropy/table/column.py [hunk 1]
```diff
@@ -504,7 +503,7 @@ def get_sortable_arrays(self):
 
 
 class BaseColumn(_ColumnGetitemShim, np.ndarray):
-    meta = MetaData(default_factory=OrderedDict)
+    meta = MetaData(default_factory=dict)
     # It's important that the metadata is an OrderedDict so that the order of
```
Comment (mhvk) [inline astropy/table/column.py:507]: Can you adjust this comment? Or perhaps just remove it?

File: astropy/table/meta.py [hunk 2]
```diff
@@ -76,7 +76,7 @@ def _construct_odict(load, node):
       >>> yaml.load('''!!omap [ foo: bar, mumble: quux, baz : gorp ]''')  # doctest: +SKIP
       OrderedDict([('foo', 'bar'), ('mumble', 'quux'), ('baz', 'gorp')])
     """
-    omap = OrderedDict()
+    omap = {}
```
Comment (mhvk) [inline astropy/table/meta.py:79]: I think the docstring needs to be adjusted here! Note that the doctest is skipped - I think that may be from a time when `yaml` was not a required dependency.

**Review by mhvk — COMMENTED** [review]
Looks good! Main comments are to update a comment and docstring.

Also, the readthedocs failures may be related to 6.1 having been branched...:

/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/16250/docs/
index_getting_started.rst:7: WARNING: toctree contains reference to document 'whatsnew/6.1' that doesn't have a title: no link will be generated

File: astropy/table/column.py (hunk 1 above)
Comment (taldcroft) [reply astropy/table/column.py]: Removed.

File: astropy/table/meta.py (hunk 2 above)
Comment (taldcroft) [reply astropy/table/meta.py:66]: I ended up removing the example entirely. It turns out that it didn't work as shown. First there were no imports so there was ambiguity about the `yaml` symbol (from astropy or pyyaml?). But more importantly it didn't show that this private function needs to be registered as a constructor in a custom Loader subclass. The comment now just refers to the astropy function where it is used and that should be fine.

**Comment by taldcroft** [comment]
@mhvk - thanks for the quick review, your comments have been addressed. I just rebased again and with any luck the RTD will pass now.

**Comment by taldcroft** [comment]
And as a reminder I prefer that this be squashed.

File: astropy/table/meta.py [hunk 3]
```diff
@@ -52,31 +52,18 @@ def items(self):
...
-    Construct OrderedDict from !!omap in yaml safe load.
+    Construct dict from !!omap in yaml safe load.
+
+    See ``get_header_from_yaml()`` for usage.
 
     Source: https://gist.github.com/weaver/317164
     License: Unspecified
 
     This is the same as SafeConstructor.construct_yaml_omap(),
     except the data type is changed to OrderedDict() and setitem is
```
Comment (mhvk) [inline astropy/table/meta.py:63]: Is this still true?

**Review by mhvk — APPROVED** [review]
I'll approve now, so you can merge when tests pass, but see my single comment.

**Comment by taldcroft** [comment]
@pllim @saimn  @astrofrog - I'm still seeing an RTD warning `/home/docs/checkouts/readthedocs.org/user_builds/astropy/checkouts/16250/docs/index_getting_started.rst:7: WARNING: toctree contains reference to document 'whatsnew/6.1' that doesn't have a title: no link will be generated
` which is causing the RTD build to fail.

I rebased on `main` on Friday. Any thoughts on fixing this?

**Comment by saimn** [comment]
Should be fixed with #16272.

**Review by nstarman — APPROVED** [review]
Cosmology looks G2G. Thanks!

**Comment by taldcroft** [comment]
I rebased on main which will hopefully fix the RTD failure.

**Pull request merged**